docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+# Verona Type System
+
+Verona is a statically typed language.
+This means all types are known at compile time and the compiler can assert that the objects and operations have the correct type at all times.
+Some types are inferred by the compiler, but there is no guessing: there's either one clear inferred type or it's an error.
+
+At a glance, Verona types are:
+ * Interfaces: types that can contain fields, methods and can be derived by other interfaces and classes.
+ * Classes: like interfaces, but classes are `final`, ie. cannot be derived further.
+ * Capabilities: behaviour guarantees, ex. `iso`, `mut`, `imm` and `readonly`.
+ * Unions: a type that can be two or more types, ex. `(A | B)`.
+ * Intersections: a type that must implement all types, ex. `(A & imm)`.
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an exclusive area of memory that only this pointer has access to in the entire program.
+ * `mut`, or `mutable`, means this object can be changed at will, which means it must be inside an area of memory you have exclusive access to, ie. traced back to an `iso` you hold the rights.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time. You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+ * `readonly` is a local capability that is used when you want to make sure an external function can't modify, but you still hold a mutable reference.
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+You can freeze an `iso` region into `imm`, but unfreeze an `imm` region.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
+
+The type system guarantees that:
+ * Structural soundness: Interfaces and classes form valid concrete types, operations and functions take in the correct (sometimes inferred) types.
+ * Capability guarantees are followed: The compiler checks operations on types, the runtime guarantees the semantics is followed.
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
```

Comment:
I don't think the third point is true, unless something has changed.  We were planning on providing contracts that had fail-fast behaviour so that array access bounds checks were safe to hoist out of loops (permitting vectorisation) and we did not need to make array accesses return a union type.

Reply:
We have three choices for arrays:

 1. Every array access must come with some proof that it's in-bounds.  For example, if we have index set types then we can check that every index comes from an index set that is derived from the array and has not been unioned with anything else (and the bounds haven't been changed without updating the index set), accesses are guaranteed to be in bounds.  This requires a lot of type system plumbing that we don't have and I've never seen a non-toy language implement this so I've no idea how usable it actually is.
 2. Every array access returns an option type.  For example, if I have `arr : Array[Foo]` then `arr(42)` is `Foo | ThingNotFound`.  This is pretty awful for the programmer because every array access has to be followed by a match expression to check if the access is in-bounds, even if the compiler can elide it.
 3. We provide contracts and say that the contract on `Array::apply(index : USize)` is that `index` is in bounds, if it's not then we have a contract violation.  This is nice for the compiler (contract violations may fast fail, if you have a loop accessing values in an array you can hoist the bounds check out of the loop), it's slightly bad for the system because any code has a capability to crash the program unless we can constrain contract violation to killing something smaller than a program.

In the absence of good choices, I thought we had decided 3 was the least bad.